### PR TITLE
BB-3592 Send email on successful instance update

### PR DESCRIPTION
**Pull request by arjunsinghy96**
Sends email to the BetaTestApplication user on successful instance redeployment of changes made by user.

**JIRA Tickets** : [BB-3592](https://tasks.opencraft.com/browse/BB-3592)

**Merge deadline**: None
**Testing instructions**:

1. Checkout to this PR
2. ssh to the vagrant box
3. run `make test.one 'registration.tests.test_api.InstanceDeploymentAPITestCase.test_send_email_on_application_redeployment'`and check all the tests pass

**Reviewers**
- [ ] @farhaanbukhsh

### CODE REVIEW

File: registration/signals.py [hunk 1]
```diff
@@ -54,3 +58,38 @@ def password_reset_token_created(
...
+    appserver = kwargs['appserver']
+    deployment_id = kwargs['deployment_id']
+
+    if deployment_id is None or appserver is None:
+        return
+
+    deployment = Deployment.objects.get(pk=deployment_id)
+
+    # Send emails for only user triggered emails
+    if deployment.type != DeploymentType.user.name:
```
Comment (farhaanbukhsh) [inline registration/signals.py:79]: What about deployment created while doing registration?  i.e  DeploymentType.registration.name

File: registration/signals.py [hunk 2]
```diff
@@ -54,3 +58,38 @@ def password_reset_token_created(
...
+        return
+
+    application = instance.betatestapplication_set.first()  # There should only be one
+    is_appserver_healthy = appserver.status.is_healthy_state
+
+    # Send email only for registered client application where application status is
+    # accepted. Applications with pending status will go through the approval signal
+    # and send welcome email.
+    if not application or application.status != BetaTestApplication.ACCEPTED:
```
Comment (farhaanbukhsh) [inline registration/signals.py:88]: Can an app server spawn if the BetaTestApplication status is pending?

File: registration/signals.py [hunk 3]
```diff
@@ -54,3 +58,38 @@ def password_reset_token_created(
...
+    if deployment_id is None or appserver is None:
+        return
+
+    deployment = Deployment.objects.get(pk=deployment_id)
+
+    # Send emails for only user triggered emails
+    if deployment.type != DeploymentType.user.name:
+        return
+
+    application = instance.betatestapplication_set.first()  # There should only be one
```
Comment (farhaanbukhsh) [inline registration/signals.py:82]: can we add a more meaningful comment?

**Review by farhaanbukhsh — CHANGES_REQUESTED** [review]
Overall it looks good, I have left few questions. :) Can we deploy this on staging and check?

File: registration/signals.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (arjunsinghy96) [reply registration/signals.py:79]: The `DeploymentType.registration` is handled already. We send a welcome email to the user after the deployment and mark the `BetaTestApplication` status as `ACCEPTED`.
In that case, an extra instance update email won't make sense.

File: registration/signals.py (hunk 2 above)
Comment (arjunsinghy96) [reply registration/signals.py:88]: The first deployment after the registration happens with `BetaTestApplication.PENDING` status (Welcome email sent on success).
This check serves as a check to determine that current deployment is a config update deployment.
In theory, appserver spawns are independent of `BetaTestApplication` status.

File: registration/signals.py (hunk 3 above)
Comment (arjunsinghy96) [reply registration/signals.py:82]: Sure. I would update it with
```
# The instance should be associated with only one BetaTestApplication by design, thus we can take the first record
```

**Review by farhaanbukhsh — APPROVED** [review]
👍 

- [x] I tested this: We have tested this PR on staging. On re-provisioning the instance an email is sent to the user.
- [x] I read through the code
- [ ] ~~I checked for accessibility issues~~
- [ ] ~~Includes documentation~~
- [ ] ~~I made sure any change in configuration variables is reflected in the corresponding client's `configuration-secure` repository.~~

**Pull request merged**